lib/src/api_analysis/summary.dart

```diff
@@ -0,0 +1,351 @@
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/analysis/results.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:collection/collection.dart';
+import 'package:pana/pana.dart';
+import 'package:path/path.dart' as path;
+
+import 'common.dart';
+import 'normalize_shape.dart';
+import 'shapes.dart';
+import 'shapes_ext.dart';
+
+Future<PackageShape> summarizePackage({
+  required PackageAnalysisContext context,
+  required String packagePath,
+  required bool normalize,
+}) async {
+  final pubspecString = context.readFile(path.join(
+    packagePath,
+    'pubspec.yaml',
+  ));
+  final pubspec = Pubspec.parseYaml(pubspecString);
+
+  if (pubspec.version == null) {
+    throw StateError(
+        'Specify a version number in `pubspec.yaml` before running analysis.');
+  }
+
+  final package = PackageShape(
+    name: pubspec.name,
+    version: pubspec.version!.toString(),
+    libraries: <LibraryShape>[],
+    getters: <GlobalPropertyShape>[],
+    setters: <GlobalPropertyShape>[],
+    functions: <FunctionShape>[],
+    classes: <ClassShape>[],
+    extensions: <ExtensionShape>[],
+    typedefs: <TypedefShape>[],
+  );
+
+  List<Annotation> summarizeAnnotations(Element element) => [
+        if (element.hasDeprecated) Annotation.deprecated,
+        if (element.hasSealed) Annotation.sealed,
+        if (element.hasVisibleForOverriding) Annotation.visibleForOverriding,
+        if (element.hasVisibleForTesting) Annotation.visibleForTesting,
+      ];
+
+  MethodShape summarizeMethod(MethodElement methodElement) {
+    return MethodShape(
+      name: methodElement.name,
+    );
+  }
+
+  PropertyShape summarizeProperty(PropertyAccessorElement property) {
+    return PropertyShape(name: property.variable.name);
+  }
+
+  NamedConstructorShape summarizeNamedConstructor(
+      ConstructorElement constructor) {
+    return NamedConstructorShape(name: constructor.name);
+  }
+
+  FunctionShape summarizeFunction(FunctionElement functionElement) {
+    return FunctionShape(
+      id: functionElement.id,
+      name: functionElement.name,
+    );
+  }
+
+  GlobalPropertyShape summarizeGlobalProperty(
+      PropertyAccessorElement property) {
+    return GlobalPropertyShape(id: property.id, name: property.variable.name);
+  }
+
+  TypedefShape summarizeTypedef(TypeAliasElement typedefElement) {
+    return TypedefShape(
+      id: typedefElement.id,
+      name: typedefElement.name,
+      targetClassId: typedefElement.aliasedType.element2 is ClassElement
+          ? typedefElement.aliasedType.element2!.id
+          : null,
+    );
+  }
+
+  ExtensionShape summarizeExtension(ExtensionElement extensionElement) {
+    // an accessor is a getter or a setter
+    final publicAccessors = extensionElement.accessors
+        .where((element) => !element.isStatic)
+        .where((element) => element.isPublic);
```

Comment:
Is turning an `Iterable` into a `List` beneficial for cases where it is accessed more than once?
Or is there also a performance penalty associated with chaining `.where()` calls?
Or both?

Reply:
> Is turning an Iterable into a List beneficial for cases where it is accessed more than once?
It will avoid calculating the condition twice - for very simple conditions calculating twice might be cheaper than allocating the list:

```
main() {
  final x = [1,2,3].where((x) {
    print(x);
    return x.isOdd;
  });
  x.toList();
  x.toList();
}
```
prints
```
1
2
3
1
2
3
```

while
```dart
main() {
  final x = [1,2,3].where((x) {
    print(x);
    return x.isOdd;
  }).toList();
  x.toList();
  x.toList();
}
```
prints:
```
1
2
3
```

> Or is there also a performance penalty associated with chaining .where() calls?

You would have to benchmark, but I very much doubt that the compiler can optimize chained `.where()` calls.

That means chained where calls has the overhead of invoking a closure (up to) twice as often. I think it is better to combine the conditions if performance is a concern.

Reply:
here https://github.com/dart-lang/pana/pull/1121/commits/775ba8411e9526619027e501fcb79d757ce13db3 I optimized iterables which are accessed more than once by converting them to lists, but I think chained `.where()` calls can be left as-is for readability, as opposed to being merged with logical operators